gym/utils/play.py

```diff
@@ -83,63 +135,30 @@ def callback(obs_t, obs_tp1, action, rew, done, info):
         If None, default key_to_action mapping for that env is used, if provided.
     """
     env.reset()
-    rendered = env.render(mode="rgb_array")
+    game = PlayableGame(env, keys_to_action, zoom)
 
-    if keys_to_action is None:
-        if hasattr(env, "get_keys_to_action"):
-            keys_to_action = env.get_keys_to_action()
-        elif hasattr(env.unwrapped, "get_keys_to_action"):
-            keys_to_action = env.unwrapped.get_keys_to_action()
-        else:
-            assert False, (
-                env.spec.id
-                + " does not have explicit key to action mapping, "
-                + "please specify one manually"
-            )
-    relevant_keys = set(sum(map(list, keys_to_action.keys()), []))
-
-    video_size = [rendered.shape[1], rendered.shape[0]]
-    if zoom is not None:
-        video_size = int(video_size[0] * zoom), int(video_size[1] * zoom)
-
-    pressed_keys = []
-    running = True
-    env_done = True
-
-    screen = pygame.display.set_mode(video_size)
+    done = True
     clock = pygame.time.Clock()
 
-    while running:
-        if env_done:
-            env_done = False
+    while game.running:
+        if done:
+            done = False
             obs = env.reset()
         else:
-            action = keys_to_action.get(tuple(sorted(pressed_keys)), 0)
+            action = keys_to_action.get(tuple(sorted(game.pressed_keys)), 0)
```

Comment:
Can we always safely assume that `0` is a good default action? (if I'm understanding this correctly)
Maybe this should be part of the specification?

In particular this seems like it'd crash with any continuous-action environments?

Reply:
Yes it is actually the default action if no relevant keys are pressed; I have the same doubt regarding ```0```, I will explore it

Reply:
@RedTachyon About this I think we can safely say that, since we can not map a continuos-action environment to discrete keyboard presses it is ok to leave ```0``` as default because ```play()``` will crash before since there is no ```keys_to_action```. (Maybe we could add an explicit message for continuos-action environment?)
An example of a keys to action mapping it is indeed this:
https://github.com/openai/gym/blob/5ae6bf9634c73ed09d73ceeb41e3491d7e791cce/gym/envs/classic_control/mountain_car.py#L240-L242